lib/streamlit/connections/snowpark_connection.py

```diff
@@ -0,0 +1,139 @@
+
+# NOTE: We won't always be able to import from snowflake.snowpark.session so need the
+# `type:ignore` comment below, but that comment will explode if `warn-unused-ignores` is
+# turned on when the package is available. Unfortunately, mypy doesn't provide a good
+# way to configure this at a per-line level :(
+# mypy: no-warn-unused-ignores
+
+import configparser
+import os
+import threading
+from collections import ChainMap
+from contextlib import contextmanager
+from datetime import timedelta
+from typing import TYPE_CHECKING, Any, Dict, Iterator, Optional, Union, cast
+
+import pandas as pd
+
+from streamlit.connections import ExperimentalBaseConnection
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.caching import cache_data
+
+if TYPE_CHECKING:
+    from snowflake.snowpark.session import Session  # type: ignore
+
+
+_REQUIRED_CONNECTION_PARAMS = {"account", "user"}
+_DEFAULT_CONNECTION_FILE = "~/.snowsql/config"
+
+
+def _load_from_snowsql_config_file() -> Dict[str, Any]:
+    """Loads the dictionary from snowsql config file."""
+    snowsql_config_file = os.path.expanduser(_DEFAULT_CONNECTION_FILE)
+    if not os.path.exists(snowsql_config_file):
+        return {}
+
+    config = configparser.ConfigParser(inline_comment_prefixes="#")
+    config.read(snowsql_config_file)
+
+    conn_params = config["connections"]
+    conn_params = {k.replace("name", ""): v.strip('"') for k, v in conn_params.items()}
+    if "db" in conn_params:
+        conn_params["database"] = conn_params["db"]
+        del conn_params["db"]
+    return conn_params
+
+
+class Snowpark(ExperimentalBaseConnection["Session"]):
+    """A thin wrapper around snowflake.snowpark.session.Session that makes it play
+    nicely with st.experimental_connection.
+
+    The Snowpark connection additionally ensures the underlying Snowpark Session can
+    only be accessed by a single thread at a time as Session object usage is *not* thread
+    safe.
+
+    NOTE: We don't expect this iteration of the Snowpark connection to be able to scale
+    well in apps with many concurrent users due to the lock contention that will occur
+    over the single underlying Session object under high load.
+    """
+
+    def __init__(self, connection_name: str, **kwargs) -> None:
+        self._lock = threading.RLock()
+
+        # Grab the lock before calling ExperimentalBaseConnection.__init__() so that we
+        # can guarantee thread safety when the parent class' constructor initializes our
+        # connection.
+        with self._lock:
+            super().__init__(connection_name, **kwargs)
+
+    # TODO(vdonato): Teach the .connect() method how to automagically connect in a SiS
+    # runtime environment.
+    def _connect(self, **kwargs) -> "Session":
+        from snowflake.snowpark.session import Session
+
+        conn_params = ChainMap(
+            kwargs,
+            self._secrets.to_dict(),
+            _load_from_snowsql_config_file(),
```

Comment:
It occurred to me yesterday that adding this to every connection could be annoying in some case where I need to maintain connections to multiple accounts to use in different contexts.

Flagging that concern here and I'm going to consider today / tomorrow AM whether there's a reasonable quick improvement or what's the best way to handle it.

Reply:
Here's the doc on named connections: https://docs.snowflake.com/en/user-guide/snowsql-start#using-named-connections

It would be nice if we either:
- check for `[connections.<name>]` in the snowsql file as higher precedence than `[connections]` (merge), OR
- check for `[connections.<name>]` in the snowsql file, IFF not found then check for default `[connections]` values

IMO either approach is fine

Reply:
Addressing in https://github.com/streamlit/streamlit/pull/6523/commits/d8aaa3e9a2e70a336765544983769ae80789d14b